### PR TITLE
Renamed insert_or_update and update from multiple_subscription_map bo…

### PR DESCRIPTION
**Pull request by kleunen**
Renamed insert_or_update and update from multiple_subscription_map both to insert and return whether element was inserted or already exists

### CODE REVIEW

**Comment by redboltz** [comment]
Could you remove subscription_map.cpp update from the PR ?
And could you rebase the PR ?

I already renamed insert_or_update with insert. 
See
https://github.com/redboltz/mqtt_cpp/pull/687/commits/13d5192acc04bd479b38ec426f3d7bc205f69b31

And
https://github.com/redboltz/mqtt_cpp/pull/688/files#diff-a35a6276aa5d36d9f80e25692e3bc78a7009bd82f326c9e54cc31c5661aedca9R158
should be remained as insert_or_update. It is for multi_index. It is actually done insert or update.

**Comment by redboltz** [comment]
Now, I'm checking CI issue on #687.

**Comment by redboltz** [comment]
It still CI processing but it seems that the problem has been fixed.
So it is ready for #688. Could you rebase #688 from sub_map_test ?

**Comment by kleunen** [comment]
It is rebased now, but I have some more renames from insert_or_update to insert than you had. Also, i renamed the remove to erase, is that OK ?

**Comment by redboltz** [comment]
I will add comments in the code.

**Comment by kleunen** [comment]
> I will add comments in the code.

Also there was a problem in the CI, i fixed this now

File: test/subscription_map.cpp [hunk 1]
```diff
@@ -156,7 +155,7 @@ BOOST_AUTO_TEST_CASE( multi_non_wc_crud ) {
     auto con1 = std::make_shared<endpoint_t>(1);
     auto con2 = std::make_shared<endpoint_t>(2);
 
-    auto insert_or_update =
+    auto insert =
```
Comment (redboltz) [inline test/subscription_map.cpp:158]: It shouldn't be replaced.
It is for multi_index container. It does actually insert or update.
See the implementation.

File: test/subscription_map.hpp [hunk 2]
```diff
@@ -337,7 +337,7 @@ class single_subscription_map
     }
 
     // Remove a value at the specified subscription path
-    void remove(MQTT_NS::string_view subscription) {
+    void erase(MQTT_NS::string_view subscription) {
```
Comment (redboltz) [inline test/subscription_map.hpp:340]: This is similar to https://en.cppreference.com/w/cpp/container/set/erase (3).
If it is easy and cheap operation, return the number of erased elements is better.

File: test/subscription_map.hpp [hunk 3]
```diff
@@ -405,20 +398,13 @@ class multiple_subscription_map
...
-            subscription_set.insert(std::move(value));
-        }
+        bool insert_result = subscription_set.insert(std::move(value)).second;
+        return std::make_pair(h, insert_result);
     }
 
     // Remove a value at the specified subscription path
     // returns the value of the removed element (if found)
-    boost::optional<Value> remove(MQTT_NS::string_view subscription, Value const& value) {
+    size_t remove(MQTT_NS::string_view subscription, Value const& value) {
```
Comment (redboltz) [inline test/subscription_map.hpp:407]: I think that the name should be `eraseP too.

File: test/subscription_map.cpp (hunk 1 above)
Comment (kleunen) [reply test/subscription_map.cpp:158]: i will fix this

File: test/subscription_map.hpp (hunk 3 above)
Comment (kleunen) [reply test/subscription_map.hpp:407]: i will fix also

File: test/subscription_map.hpp (hunk 2 above)
Comment (kleunen) [reply test/subscription_map.hpp:340]: Yes, very cheap, just returning the result of the erase from the underlying set.

Comment (kleunen) [reply test/subscription_map.hpp:340]: ah, I made a mistake in the erase.

Comment (kleunen) [reply test/subscription_map.hpp:340]: The unit test should include the subscription is removed when it is empty in multiple_subscription_map, when all the values are removed from a subscription, the subscription can be removed completely from the map. Also the reference counting should be checked in the unit test. In my test I included checking that when all subscriptions are removed, you only have the root element left.

**Pull request merged**